### PR TITLE
Add a lot of resilience-related annotations to stdlib to improve the performance of the resilient stdlib

### PR DESCRIPTION
**Pull request by swiftix**

### CODE REVIEW

**Comment by swiftix** [comment]
@airspeedswift @dabrahams This is the initial set of annotations, which results in a pretty good performance for many of our benchmarks. Please have a look to see if I touched something that should not be made @_inlineable, @_versioned or @_fixed_layout.

The PR does not pass testing yet, because there are about 10 tests that need to be updated.

**Comment by swiftix** [comment]
@slavapestov If you take this PR and revert the last commit: https://github.com/apple/swift/pull/7350/commits/17ed6bae465cf48bfce89ba1f44991e5eaa243fa, then you get a compiler crasher. Could you have a look? I hope it is easy to fix, just like the previous resilience related bugs I've reported.

**Comment by slavapestov** [comment]
I'll take a look, thanks @swiftix.

**Comment by slavapestov** [comment]
@swiftix Designated initializers on classes cannot be @_inlineable for the same reason as designated initializers on non-fixed-layout value types. Is it important for these initializers to be inlineable? If not, I'll add a diagnostic prohibiting them. If so, we should discuss some options.

**Comment by swiftix** [comment]
@slavapestov Thanks for improving the diagnostics. One minor annoying thing I noticed: when the compiler produces a diagnostics that @_inlinable cannot be used with stored properties, it does not show the name of the stored property. Could you add the name of the properties to the message? it is very annoying to search by the line number in gyb generated Swift files ;-) Searching by name in the original gyb/swift files is much easier.

**Comment by dabrahams** [comment]
If the line-directive tool is not repointing those errors at the original .gyb files then something is very broken.

**Comment by swiftix** [comment]
@airspeedswift @dabrahams Ping! I rebased on ToT. Please have a look. In particular, check if I put @_fixed_layout or @_versioned on types, where you do not want to expose the representation or any details about it.

**Comment by swiftix** [comment]
@swift-ci please smoke test

File: benchmark/single-source/PopFrontGeneric.swift [hunk 1]
```diff
@@ -17,6 +17,7 @@ let arrayCount = 1024
 
 // This test case exposes rdar://17440222 which caused rdar://17974483 (popFront
 // being really slow).
+@_versioned
```
Comment (slavapestov) [inline benchmark/single-source/PopFrontGeneric.swift]: Why did the benchmark change?

File: stdlib/private/StdlibUnittestFoundationExtras/StdlibUnittestFoundationExtras.swift [hunk 2]
```diff
@@ -75,6 +75,8 @@ public func autoreleasepoolIfUnoptimizedReturnAutoreleased(
 #endif
 }
 
+@_inlineable
+@_versioned
```
Comment (slavapestov) [inline stdlib/private/StdlibUnittestFoundationExtras/StdlibUnittestFoundationExtras.swift]: This function has no body, it should not be @_inlineable (I'll add a diagnostic for that)

File: stdlib/public/core/Arrays.swift.gyb [hunk 3]
```diff
@@ -451,9 +456,9 @@ if True:
 }%
 
 ${SelfDocComment}
-% if Self != 'ArraySlice':
+//% if Self != 'ArraySlice':
```
Comment (slavapestov) [inline stdlib/public/core/Arrays.swift.gyb]: Just remove the gybbing here?

File: stdlib/public/core/Arrays.swift.gyb [hunk 4]
```diff
@@ -473,11 +478,15 @@ public struct ${Self}<Element>
   /// For an instance of `${Self}`, `startIndex` is always zero. If the array
   /// is empty, `startIndex` is equal to `endIndex`.
 %end
+  @_inlineable
   public var startIndex: Int {
 %if Self == 'ArraySlice':
     return _buffer.startIndex
 %else:
-    return 0
+    @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Arrays.swift.gyb]: Oops, the @_inlineable should apply to the getter too. Can you revert this and I'll make sure it works?

File: stdlib/public/core/Arrays.swift.gyb [hunk 5]
```diff
@@ -994,6 +1046,8 @@ extension ${Self} : RangeReplaceableCollection, _ArrayProtocol {
   ///     emptyArray = []
   ///     print(emptyArray.isEmpty)
   ///     // Prints "true"
+  // @inline(__always)
```
Comment (slavapestov) [inline stdlib/public/core/Arrays.swift.gyb]: Remove this line?

File: stdlib/public/core/Map.swift.gyb [hunk 6]
```diff
@@ -31,20 +32,32 @@ public struct LazyMapIterator<
   ///
   /// - Precondition: `next()` has not been applied to a copy of `self`
   ///   since the copy was made.
+  @_inlineable
   public mutating func next() -> Element? {
     return _base.next().map(_transform)
   }
 
+  @_inlineable
   public var base: Base { return _base }
```
Comment (slavapestov) [inline stdlib/public/core/Map.swift.gyb]: I wonder if this could be private(set) instead

File: stdlib/public/core/Map.swift.gyb [hunk 7]
```diff
@@ -157,49 +187,61 @@ public struct ${Self}<
...
   }
 
+  @_inlineable
   public var underestimatedCount: Int {
     return _base.underestimatedCount
   }
 
   /// Create an instance with elements `transform(x)` for each element
   /// `x` of base.
+  //@_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Map.swift.gyb]: Why wasn't this one inlineable?

File: stdlib/public/core/Mirror.swift [hunk 8]
```diff
@@ -875,6 +876,7 @@ extension String {
   ///     // Prints "(21, 30)"
   ///
   /// - SeeAlso: `String.init<Subject>(reflecting: Subject)`
+  @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Mirror.swift]: Mirrors are not performance-critical, I think you should remove this

File: stdlib/public/core/Optional.swift [hunk 9]
```diff
@@ -670,10 +676,13 @@ public func ?? <T>(optional: T?, defaultValue: @autoclosure () throws -> T?)
 }
 
 extension Optional {
+
+  @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Optional.swift]: These are not available and should not be inlineable

File: stdlib/public/core/OutputStream.swift [hunk 10]
```diff
@@ -386,13 +387,17 @@ internal func _print_unlocked<T, TargetStream : TextOutputStream>(
 ///
 /// This function is forbidden from being inlined because when building the
 /// standard library inlining makes us drop the special semantics.
+@_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/OutputStream.swift]: This is inline(never) so it doesn't have to be inlineable either -- you'll just end up emitting useless copies of this into client binaries

File: stdlib/public/core/Range.swift.gyb [hunk 11]
```diff
@@ -108,33 +109,43 @@ public struct CountableRange<Bound> : RandomAccessCollection
   public typealias IndexDistance = Bound.Stride
 
   public var startIndex: Index {
-    return lowerBound
+    @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Range.swift.gyb]: Redundant 'get {'

File: stdlib/public/core/Range.swift.gyb [hunk 12]
```diff
@@ -173,6 +184,7 @@ public struct CountableRange<Bound> : RandomAccessCollection
   /// (`..<`) to form `CountableRange` instances is preferred.
   ///
   /// - Parameter bounds: A tuple of the lower and upper bounds of the range.
+  @inline(__always)
```
Comment (slavapestov) [inline stdlib/public/core/Range.swift.gyb]: Why not leave the decision up to the optimizer?

File: stdlib/public/core/RangeReplaceableCollection.swift.gyb [hunk 13]
```diff
@@ -1207,6 +1238,7 @@ public func += <
 public typealias RangeReplaceableCollectionType = RangeReplaceableCollection
 
 extension RangeReplaceableCollection {
+  @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/RangeReplaceableCollection.swift.gyb]: Unavailable methods don't need to be inlineable

File: stdlib/public/core/Repeat.swift [hunk 14]
```diff
@@ -105,6 +112,7 @@ public func repeatElement<T>(_ element: T, count n: Int) -> Repeated<T> {
 public struct Repeat<Element> {}
 
 extension Repeated {
+  @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Repeat.swift]: Unavailable methods don't need to be inlineable

File: stdlib/public/core/OutputStream.swift (hunk 10 above)
Comment (swiftix) [inline stdlib/public/core/OutputStream.swift]: May be we want to specialize it?

File: stdlib/public/core/Sequence.swift [hunk 15]
```diff
@@ -1081,6 +1132,7 @@ extension Sequence {
   ///   element is a match.
   /// - Returns: The first element of the sequence that satisfies `predicate`,
   ///   or `nil` if there is no element that satisfies `predicate`.
+  // @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Sequence.swift]: Why wasn't this inlineable?

File: stdlib/public/core/Sequence.swift [hunk 16]
```diff
@@ -1429,16 +1494,19 @@ public typealias GeneratorType = IteratorProtocol
 public typealias SequenceType = Sequence
 
 extension Sequence {
+  @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Sequence.swift]: Unavailable methods don't need to be inlineable

File: stdlib/public/core/SwiftNativeNSArray.swift [hunk 17]
```diff
@@ -225,20 +256,32 @@ extension _SwiftNativeNSArrayWithContiguousStorage : _NSArrayCore {
...
 /// Base class of the heap buffer backing arrays.  
+@_versioned
+@_fixed_layout
 internal class _ContiguousArrayStorageBase
   : _SwiftNativeNSArrayWithContiguousStorage {
 
+  @_versioned
   final var countAndCapacity: _ArrayBody
 
+  @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/SwiftNativeNSArray.swift]: These constructors just throw errors, don't have to be inlinable

File: stdlib/public/core/SwiftNativeNSArray.swift [hunk 18]
```diff
@@ -225,20 +256,32 @@ extension _SwiftNativeNSArrayWithContiguousStorage : _NSArrayCore {
...
+  @_versioned
```
Comment (slavapestov) [inline stdlib/public/core/SwiftNativeNSArray.swift]: I don't think this should be versioned either, it says doNotCallMe :-)

File: stdlib/public/core/SwiftNativeNSArray.swift [hunk 19]
```diff
@@ -124,26 +139,34 @@ extension _SwiftNativeNSArrayWithContiguousStorage : _NSArrayCore {
 ///
 /// Ideally instances of this class would be allocated in-line in the
 /// buffers used for Array storage.
+@_fixed_layout
```
Comment (slavapestov) [inline stdlib/public/core/SwiftNativeNSArray.swift]: We should measure the overload of having these classes not be fixed layout (which means the ctor cannot be inlined). Maybe inlining the allocating entry point (which should always be safe) will make up for it :)

File: stdlib/public/core/OutputStream.swift (hunk 10 above)
Comment (slavapestov) [inline stdlib/public/core/OutputStream.swift]: I see. In practice does specializing here make a difference? Perhaps just having some pre-specializations is sufficient.

**Comment by slavapestov** [comment]
@swiftix Does reverting this commit here still crash the compiler? https://github.com/apple/swift/pull/7350/commits/5252041634e5aeb10587940407a9b537b5c222c3

**Comment by slavapestov** [comment]
@swiftix It looks like '@_inlineable var foo: Int { return 3 }' and '@_inlineable var foo: Int { get { return 3 } }' both work. I think you can remove the unnecessary 'get {}'s from this PR now.

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please benchmark

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by dabrahams** [comment]
What I hate about this: having merged this into my branch I am getting piles of errors like

```
error: instance method '_forceCreateUniqueMutableBuffer(countForNewBuffer:minNewCapacity:)' is internal and cannot be referenced from an '@_inlineable' function
```

and I don't know the rules.  @slavapestov @swiftix HALP!

**Comment by swiftix** [comment]
@dabrahams An `@_inlineable` function can only reference:
- public functions
- @_inlineable functions
- @_versioned functions.

So, in your case you most likely want to add `@_versioned` to your internal method. 

If you think that it makes sense to make this internal method also potentially available for inlining (i.e. you want to serialize its SIL body), you can also add `@_inlineable` in addition to `@_versioned`.

**Comment by dabrahams** [comment]
@swiftix Thanks; I think I got past the errors eventually.  Not sure I did it "right", but the merge is 435a67afd822a7457c6eec0c4defca2d5d7098cc if you want to look.

**Comment by palimondo** [comment]
@swiftix I’d love to, but I can not get the swift-source to compile on my machine… any pointers to where I might get help?

**Comment by swiftix** [comment]
@palimondo You could ask questions on the swift-dev@swift.org mailing list.

**Pull request merged**